### PR TITLE
Modified Invoke-AtomicRedTeam functions and README

### PR DESCRIPTION
**Pull request by MSAdministrator**
**Details:**

I took some liberties here and updated the `Invoke-AtomicRedTeam` functions to align to some best practices.  I added:

- Modified logic for installing dependencies
- Modified logic in loops a bit
- Added ability to use CommonParameters like WhatIf, Confirm, Verbose, Debug, and InformationAction for controlling actions and output
- Added Progress bars and status output
- Updated the README to reflect changes

**Testing:**

I tested this on a Windows 10 x64 VM.  I know the minimum PowerShell version this previously supported was PowerShell 3.0 or higher.  I kept the same compatibility.

**Associated Issues:**

### CODE REVIEW

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 1]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
+    [CmdletBinding(DefaultParameterSetName = 'technique',
+        SupportsShouldProcess = $true,
+        PositionalBinding = $false,
+        ConfirmImpact = 'Medium')]
     Param(
+        [Parameter(Mandatory = $true,
+            Position = 0,
+            ValueFromPipelineByPropertyName = $true,
+            ParameterSetName = 'technique')]
+        [ValidateNotNull()]
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is extraneous unless I'm remembering how `ValidateNotNullOrEmpty` work.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 2]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
+
+    Process 
+    {
+        Write-Verbose -Message 'Attempting to convert files from yaml'
         foreach ($File in $Path) {
-            $parsedYaml = (ConvertFrom-Yaml (Get-Content $File -Raw ))
-            Write-Output $parsedYaml
+            if ($pscmdlet.ShouldProcess($File, 'Converting yaml file')) {
+                Write-Verbose -Message "Converting $File from Yaml"]
+                $parsedYaml = (ConvertFrom-Yaml (Get-Content $File -Raw))
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: Nit: Unnecessary parentheses here.

(I know it was in the original code.)

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 3]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
+    [CmdletBinding(DefaultParameterSetName = 'technique',
+        SupportsShouldProcess = $true,
+        PositionalBinding = $false,
+        ConfirmImpact = 'Medium')]
     Param(
+        [Parameter(Mandatory = $true,
+            Position = 0,
+            ValueFromPipelineByPropertyName = $true,
+            ParameterSetName = 'technique')]
+        [ValidateNotNull()]
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: Unnecessary unless there's a subtlety I'm forgetting about this attribute.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 4]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
+            $techniqueCount++
+            Write-Progress -Activity "Running $($Technique.display_name.ToString()) Technique" -Status 'Progress:' -PercentComplete ($techniqueCount / ($AtomicTechnique).Count * 100)
+            Write-Debug -Message "Gathering tests for Technique $Technique"
 
-            foreach ($Test in $AtomicTest) {
-                #Only Process Windows Tests For Now
+            $testCount = 0
+            foreach ($Test in $Technique.atomic_tests) {
+                $testCount++
+                Write-Progress -Activity 'Running Atomic Tests' -Status 'Progress:' -PercentComplete ($testCount / ($Technique.atomic_tests).Count * 100)
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This line's getting a bit long. I'd do the math as its own variable in a line above.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 5]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
-    BEGIN {}
+    BEGIN { } # Intentionally left blank and can be removed
     PROCESS {
+        Write-Verbose -Message 'Attempting to run Atomic Techniques'
+
+        $techniqueCount = 0
         foreach ($Technique in $AtomicTechnique) {
-            $AtomicTest = $Technique.atomic_tests
+            $techniqueCount++
+            Write-Progress -Activity "Running $($Technique.display_name.ToString()) Technique" -Status 'Progress:' -PercentComplete ($techniqueCount / ($AtomicTechnique).Count * 100)
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This line's getting a bit long. I'd do the math in its own variable a line above.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 6]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
+            Write-Debug -Message "Gathering tests for Technique $Technique"
 
-            foreach ($Test in $AtomicTest) {
-                #Only Process Windows Tests For Now
+            $testCount = 0
+            foreach ($Test in $Technique.atomic_tests) {
+                $testCount++
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: We're using inconsistent variable capitalization in this file with the addition of camelCase on top of the PascalCase being used elsewhere. I think we should be consistent.

I don't have strong feelings either way though in my experience, I've usually seen:

- PascalCase for Global variables
- camelCase for local variables

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 7]
```diff
@@ -33,75 +30,127 @@ Github repo: https://github.com/redcanaryco/atomic-red-team
...
-                    $Technique.display_name.ToString(), $Technique.attack_technique.ToString())
-                Write-Host $Test.name.ToString()
-                Write-Host $Test.description.ToString()
+                Write-Information -MessageData ("[********BEGIN TEST*******]`n" +
+                    $Technique.display_name.ToString(), $Technique.attack_technique.ToString()) -Tags 'Details'
+                
+                Write-Information -MessageData $Test.name.ToString() -Tags 'Details'
+                Write-Information -MessageData $Test.description.ToString() -Tags 'Details'
 
+                Write-Debug -Message 'Gathering final executation command'
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: Type: execution?

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 [hunk 8]
```diff
@@ -111,37 +160,41 @@ function Invoke-AtomicTest {
...
-            }
-
-            Write-Host "[!!!!!!!!END TEST!!!!!!!]`n`n" -Foreground Yellow
-        }
-    }
-    END { }
+                    Write-Verbose -Message 'Invoking Atomic Tests using defined executor'
+                    if ($pscmdlet.ShouldProcess(($Test.name.ToString()), 'Execute Atomic Test')) {
+                        switch ($Test.executor.name) {
+
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: Nit: extraneous space.

**Comment by 2xyo** [comment]
It could be useful to have the ability to specify the path of the `powershell-yaml` module as you maybe don't want to install this dependency as you wish to keep a small footprint.

**Comment by 2xyo** [comment]
It also could be useful to [check the Powershell version ](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-3.0#-version-nn)as the script doesn't work with Powershell v2.

**Comment by zacbrown** [comment]
@2xyo - does the script need to support this or can the person executing it simply modify `$PSModulePath` for the current powershell session?

That seems like a power user scenario that adds complexity to the script where a well established pattern might be simpler. 

What do you think?

**Comment by 2xyo** [comment]
@zacbrown The choice of to not install dependencies is part of my use case (keep a minimal footprint). Unfortunately, I can't say if it's could be really useful for the other users  :confused:
If this option looks "too much", having the bypass with `$PSModulePath` in the documentation or in the error message would be perfect.

**Comment by MSAdministrator** [comment]
Hey everyone, its been a busy week at work for me but I plan on responding/reading all the comments later this evening. Just an FYI - and I’m not ingnoring anyone. :)

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 1 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: You are correct!  My mistake!  I'll remove that.

Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 2 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 3 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 4 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 5 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 6 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 7 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1 (hunk 8 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.ps1]: This is resolved.

**Comment by MSAdministrator** [comment]
I have taken some more liberties and converted this to a PowerShell Script Module per community standards.  Let me know what you think.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    ModuleVersion     = '1.0.0.0'
+
+    # Supported PSEditions
+    CompatiblePSEditions = @('Desktop')
+
+    # ID used to uniquely identify this module
+    GUID              = '8f492621-18f8-432e-9532-b1d54d3e90bd'
+
+    # Author of this module
+    Author            = 'Casey Smith @subTee'
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: While Casey is the original author, I think he would agree that you should be included given the significant improvements you've provided. Please add yourself if you're comfortable with that :).

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    # List of all files packaged with this module
+    # FileList = @()
+
+    # Private data to pass to the module specified in RootModule/ModuleToProcess. This may also contain a PSData hashtable with additional module metadata used by PowerShell.
+    PrivateData       = @{
+
+        PSData = @{
+
+            # Tags applied to this module. These help with module discovery in online galleries.
+            Tags = @('Red Canary', 'Atomic', 'Red Team', 'MITRE', 'ATT&CK' )
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: Might be worthwhile to add "ART"? Thoughts @caseysmithrc?

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+            # A URL to the license for this module.
+            # LicenseUri = 'http://opensource.org/licenses/MIT'
```
Comment (zacbrown) [inline execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: You can uncomment this. We are using MIT license for ART - if you want to reference a specific file within the ART repo, this should be sufficient: https://github.com/redcanaryco/atomic-red-team/blob/master/LICENSE.txt

**Comment by MSAdministrator** [comment]
@zacbrown I will make the suggested changes (Thanks!), but I wanted to ask if your team has a `NuGetApiKey` for the PowerShell Gallery?  I see you're using circleci.  I typically use AppVeyor for CI and once a branch has passed it will increment the build number in the manifest and publish to the PowerShell Gallery.

I will add some basic Pester tests for structure, but it may have to be integrated into your circleci yaml.  Here's the newest appveyor.yml i've been using to do this: https://github.com/MSAdministrator/PSNamedPipe/blob/master/appveyor.yml

**Comment by zacbrown** [comment]
@MSAdministrator - We don't as far as I know. I've filed issue #345 to do this work as a follow up fix.

I think we're fine to move forward with your changes with the fixes for my comments and we'll work toward the powershellgallery.com publishing in a separate change.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 (hunk 9 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 (hunk 10 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: This is resolved.

File: execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1 (hunk 11 above)
Comment (MSAdministrator) [reply execution-frameworks/Invoke-AtomicRedTeam/Invoke-AtomicRedTeam.psd1]: This is resolved.

**Pull request merged**